File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_a1/organic7_binary_search.java
/*
 * Binary Search - Submission by Ramon
 * Approach: Object-oriented design - a Searcher class wraps the sorted
 * data and exposes a find() method.
 */

class Searcher {
    private int[] data;

    public Searcher(int[] data) {
        this.data = data;
    }

    public int find(int target) {
        int lo = 0;
        int hi = this.data.length - 1;
        while (lo <= hi) {
            int mid = (lo + hi) / 2;
            int value = this.data[mid];
            if (value == target) {
                return mid;
            } else if (value < target) {
                lo = mid + 1;
            } else {
                hi = mid - 1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] arr = {2, 5, 8, 12, 16, 23, 38, 45, 56, 72, 91};
        int target = 23;
        Searcher searcher = new Searcher(arr);
        System.out.println("Class-based result: " + searcher.find(target));
    }
}

File: server/datasets/controlled_benchmarks/python/multiple_files/multiple_a1/organic7_binary_search.py
"""
Binary Search Suite - Submission by Ramon
Approach: Object-oriented BinarySearchEngine class encapsulating dataset,
caching results, and providing analytical summary methods.
"""


class BinarySearchEngine:
    def __init__(self, dataset):
        self.data = dataset
        self.length = len(dataset)
        self.query_history = []

    def find(self, target):
        lo, hi = 0, self.length - 1
        comps = 0
        while lo <= hi:
            comps += 1
            mid = (lo + hi) // 2
            val = self.data[mid]
            if val == target:
                self.query_history.append((target, mid, comps))
                return mid
            elif val < target:
                lo = mid + 1
            else:
                hi = mid - 1
        self.query_history.append((target, -1, comps))
        return -1

    def count_matches(self, target):
        pos = self.find(target)
        if pos == -1:
            return 0
        left = pos
        while left > 0 and self.data[left - 1] == target:
            left -= 1
        right = pos
        while right < self.length - 1 and self.data[right + 1] == target:
            right += 1
        return (right - left) + 1

    def summary(self):
        return {"total_queries": len(self.query_history), "history": self.query_history}


if __name__ == "__main__":
    arr = [2, 5, 8, 12, 12, 12, 23, 38, 45, 56, 72, 91]
    engine = BinarySearchEngine(arr)
    print("Index:", engine.find(12))
    print("Matches:", engine.count_matches(12))
    print("Summary:", engine.summary())
